src/eetcd_election.erl

```diff
@@ -8,11 +8,22 @@
 -export([observe/3, observe_stream/2]).
 
 -export_type([campaign_ctx/0, observe_ctx/0]).
--type observe_ctx() :: #{leader => map(), http2_pid => pid(), monitor_ref => reference(), stream_ref => reference()}.
--type campaign_ctx() :: #{campaign => map()|'waiting_campaign_response', http2_pid => pid(), monitor_ref => reference(), stream_ref => reference()}.
+-type leader_key() :: router_pb:'Etcd.LeaderKey'().
+-type observe_ctx() :: #{
+    leader => map() | election_no_leader,
+    http2_pid => pid(),
+    monitor_ref => reference(),
+    stream_ref => reference()
+}.
+-type campaign_ctx() :: #{
+    campaign => leader_key() | waiting_campaign_response,
+    http2_pid => pid() | undefined,
+    monitor_ref => reference() | undefined,
+    stream_ref => reference() | undefined
```

Comment:
The last three fields are already optionally associated. It's better to remove them instead of setting as `undefined` in L147-149.